ledger-tool/src/ledger_utils.rs

```diff
@@ -294,27 +293,41 @@ pub fn load_and_process_ledger(
         None,
     );
 
-    let (transaction_status_sender, transaction_status_service) = if transaction_notifier.is_some()
-    {
-        let (transaction_status_sender, transaction_status_receiver) = unbounded();
-        let transaction_status_service = TransactionStatusService::new(
-            transaction_status_receiver,
-            Arc::default(),
-            false,
-            transaction_notifier,
-            blockstore.clone(),
-            false,
-            exit.clone(),
-        );
-        (
-            Some(TransactionStatusSender {
-                sender: transaction_status_sender,
-            }),
-            Some(transaction_status_service),
-        )
-    } else {
-        (None, None)
-    };
+    let enable_rpc_transaction_history = arg_matches.is_present("enable_rpc_transaction_history");
+
+    let (transaction_status_sender, transaction_status_service) =
+        if geyser_plugin_active || enable_rpc_transaction_history {
+            // Need Primary (R/W) access to insert transaction data
+            let tss_blockstore = if enable_rpc_transaction_history {
+                Arc::new(open_blockstore(
+                    blockstore.ledger_path(),
+                    AccessType::Primary,
+                    None,
+                    false,
+                ))
```

Comment:
I'm confused here. Now we have 2 blockstores, with the same path.
I'm really not familiar with blockstore internals, but it feels like this probably introduces race conditions.

Based on comment above, is the only reason we need a new one here because the access type must be primary? Or maybe I am missing some additional context.

Reply:
A little context on ` Blockstore` access types:
https://github.com/solana-labs/solana/blob/8273b103d264e0c75d3d180f96925d91b5a64449/ledger/src/blockstore_options.rs#L31-L40

> I'm confused here. Now we have 2 blockstores, with the same path.
> I'm really not familiar with blockstore internals, but it feels like this probably introduces race conditions.

`Blockstore` (and the underlying `rocksdb` db) are safe to have multiple accessors. This is really no different than having the validator running with `Primary` access and running `solana-ledger-tool` commands with `Secondary`. Only one instance can have `Primary` (r/w) access at a time (this is enforced at `rocksdb` level so it prevents us from doing stupid stuff too ha).

> Based on comment above, is the only reason we need a new one here because the access type must be primary? Or maybe I am missing some additional context.

Yes, a bit ago, we pushed a change in to make all commands run with `Secondary` access (read only), only a few commands really need `Primary` (r/w) access. `verify` specifically only needs `Secondary` as it is only reading the blocks and replaying them.

That being said, the new flag that I piped through specifically allows writing to the `Blockstore`, so we need `Primary` access to allow it to actually push the tx data. I opened a new `Blockstore` instance only for this service to restrict the Primary access to the narrowest scope possible

Reply:
> verify specifically only needs Secondary as it is only reading the blocks and replaying them.

Is this going to conflict with skipping snapshot unarchive: #31921 (renamed at some point to `use_snapshot_archives_at_startup`)

If this option is passed, `verify` will use `PrimaryForMaintainence`. Is that going to cause errors?

Reply:
>  Is that going to cause errors?

Yes, the second attempt to get primary access will not succeed (in this case, the transaction status service one).

On a more general note -  I commented on that PR when I first saw it but didn't follow along / actually review. The pre-existing logic of gating accounts/snapshots paths based on access type is hacky/conflating separate ideas.
- Access type is purely a `Blockstore` thing that refers to whether you are granted read+write or read-only permissions
- We used this access type to then gate whether we'd use the usual (whatever is passed to `solana-validator`) paths or `.ledger-tool` paths; this was a lazy way to prevent `ledger-tool` from accidentally obliterating working stuff from `solana-validator`. The main case here would be wanting to put `accounts` in `tmpfs` to get that speedup when possible (ie when `solana-validator` wasn't already using that space)

With that context stated, I think:
- This PR does the proper thing
    - Things will only fail if both flags are being used
- We should untangle blockstore access type from accounts/snapshot path selection.

Deriving paths from access type was always a hack, and that code is in need of an update. This is a known thing that has never been prioritized but with that PR you mentioned above, I think it is now time to fix it. I'll take a crack at it and include you and Brooks on the review.

Reply:
> On a more general note - I commented on that PR when I first saw it but didn't follow along / actually review. The pre-existing logic of gating accounts/snapshots paths based on access type is hacky/conflating separate ideas.

Ha yeah, I don't think you'll find argument from either me nor @brooksprumo on that.

> This is a known thing that has never been prioritized but with that PR you mentioned above, I think it is now time to fix it. I'll take a crack at it and include you and Brooks on the review.

👑 awesome!